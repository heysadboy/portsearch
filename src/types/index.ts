import combineReducers from '../redux/reducers';

export interface IPort {
    code: string,
    name: string
}

export interface IMarketRate {
    day: string,
    mean: number | null,
    low: number | null,
    high: number | null
}

export interface IPointData {
    day: string,
    value: number | null
}

export interface ILineData {
    name: string,
    color: string,
    values: IPointData[]
}

export interface IGraphData {
    [key: string]: ILineData
}

export interface IPortAction {
    type: EPortActionType
    payload: IPort[]
}

export interface IMarketRateAction {
    type: EMarketRateActionType
    payload: IMarketRate[]
}

export enum EPortActionType {
    get_ports = "GET_PORTS"
}

export enum EMarketRateActionType {
    get_ports = "GET_MARKET_RATES"
}

export type AppState = ReturnType<typeof combineReducers>;
