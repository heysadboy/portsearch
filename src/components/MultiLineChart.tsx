import { FC, useEffect } from "react";
import * as d3 from "d3";
import { ILineData } from "../types";

interface IMultiLineChartProp {
    data: ILineData[],
    domain: any
}

const MultiLineChart: FC<IMultiLineChartProp> = ({ data, domain }) => {
    
    useEffect(() => {
        d3.select("#chart").selectAll("*").remove();

        if (data.length > 0 && data[0]) {
            console.log(data);

            var margin = { top: 2, right: 30, bottom: 30, left: 60 },
                width = 800 - margin.left - margin.right,
                height = 400 - margin.top - margin.bottom;

            var svg = d3.select("#chart").append("svg")
                .attr("width", width + margin.left + margin.right)
                .attr("height", height + margin.top + margin.bottom)
                .append("g")
                .attr("transform",
                    "translate(" + margin.left + "," + margin.top + ")");

            // Adding X axis
            var x = d3.scaleTime()
                .domain(d3.extent(data[0].values, function (d: any) { return d.day; }) as Iterable<Date>)
                .range([0, width]);
            svg.append("g")
                .attr("transform", "translate(0," + height + ")")
                .call(d3.axisBottom(x).ticks(5));


            // Adding Y axis
            var y = d3.scaleLinear()
                .domain([0, domain.yMaxValue + 1000] as Iterable<number>)
                .range([height, 0]);
            svg.append("g")
                .call(d3.axisLeft(y));

            // Adding lines
            svg.selectAll(".line")
                .data(data)
                .enter()
                .append("path")
                .attr("fill", "none")
                .attr("stroke", function (d) { return d.color })
                .attr("stroke-width", 1.5)
                .attr("d", function (d: any) {
                    return d3.line()
                        .x(function (d: any) { return x(d.day); })
                        .y(function (d: any) { return y(+d.value); })
                        (d.values)
                })

        }
    }, [data, domain.yMaxValue])

    return (<div id="chart"></div>)
};

export default MultiLineChart;