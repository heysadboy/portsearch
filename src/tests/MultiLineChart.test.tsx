import TestRenderer from "react-test-renderer";
import MultiLineChart from "../components/MultiLineChart";
import { ILineData } from "../types";

const data: ILineData[] = [
    {
        name: "low",
        color: "#D50000",
        values: [
            {
                day: new Date("2021-01-01"),
                value: 1037,
            },
            {
                day: new Date("2021-01-02"),
                value: 1037,
            },
            {
                day: new Date("2021-01-03"),
                value: 1007,
            }
        ]
    }
]

test("MultiLineChart Component Testing.", () => {
    const domain = {

        yMinValue: 1007,
        yMaxValue: 1037
    }

    const component = TestRenderer.create(
        <MultiLineChart data={data} domain={domain} />
    );

    let tree: any = component.toJSON();
    expect(tree).toMatchSnapshot();
});