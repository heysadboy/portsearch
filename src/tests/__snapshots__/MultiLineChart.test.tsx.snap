// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MultiLineChart Component Testing. 1`] = `
<div
  id="chart"
/>
`;
